### PR TITLE
Fix draw order problems with south walk

### PR DESCRIPTION
**Pull request by kphoenix137**
This resolves a problem with draw order for players and monsters passing through doorways, caused by https://github.com/diasurgical/devilutionX/pull/7043. It's possible there are other situations where draw order was a problem which this should also resolve.

This PR is step 1; step 2 will be fixing draw order problems with sideways walk.

### CODE REVIEW

File: Source/engine/render/scrollrt.cpp [hunk 1]
```diff
@@ -664,17 +664,8 @@ void DrawMonsterHelper(const Surface &out, Point tilePosition, Point targetBuffe
...
-	if (monster.isWalking()) {
-		bool isSideWalkingToLeft = monster.mode == MonsterMode::MoveSideways && monster.direction == Direction::West;
-		if (isNegativeMonster && !isSideWalkingToLeft)
-			return;
-		if (!isNegativeMonster && isSideWalkingToLeft)
-			return;
-		if (isSideWalkingToLeft)
-			offset -= Displacement { 64, 0 };
-	} else if (isNegativeMonster) {
+	if (isNegativeMonster)
```
Comment (AJenbo) [inline Source/engine/render/scrollrt.cpp:667]: This is now always done. Please move it to the very top of the function and remove the other case.
```cpp
if (mi < 0)
		return;
```

File: Source/engine/render/scrollrt.cpp [hunk 2]
```diff
@@ -747,13 +738,62 @@ void DrawDungeon(const Surface &out, Point tilePosition, Point targetBufferPosit
...
+			DrawPlayerHelper(out, *player, tempTilePosition, tempTargetBufferPosition); // BUGFIX: Player sprite gets cut off walking in front of certain corners in caves
+		}
 	}
-	if (dMonster[tilePosition.x][tilePosition.y] != 0) {
-		DrawMonsterHelper(out, tilePosition, targetBufferPosition);
+
+	Monster *monster = FindMonsterAtPosition(tilePosition);
+	if (monster != nullptr) {
+		// If sprite is moving southwards, we want to draw it offset from the tile it's moving to, so we need negative ID
+		int8_t monsterId = (monster->mode != MonsterMode::MoveSouthwards) ? monster->getId() + 1 : -monster->getId() - 1;
```
Comment (AJenbo) [inline Source/engine/render/scrollrt.cpp:772]:
```suggestion
		int8_t monsterId = static_cast<int8_t>(monster->getId() + 1);
		if (monster->mode == MonsterMode::MoveSouthwards)
			monsterId = -monsterId;
```

File: Source/engine/render/scrollrt.cpp [hunk 3]
```diff
@@ -747,13 +738,62 @@ void DrawDungeon(const Surface &out, Point tilePosition, Point targetBufferPosit
...
+				case Direction::South:
+					tempTargetBufferPosition += { 0, -TILE_HEIGHT };
+					break;
+				case Direction::SouthEast:
+					tempTargetBufferPosition += { -TILE_WIDTH / 2, -TILE_HEIGHT / 2 };
+					break;
+				}
+				tempTilePosition += Opposite(player->_pdir);
+			}
+			DrawPlayerHelper(out, *player, tempTilePosition, tempTargetBufferPosition); // BUGFIX: Player sprite gets cut off walking in front of certain corners in caves
```
Comment (AJenbo) [inline Source/engine/render/scrollrt.cpp:765]: Is this new since all the s-walk changes, or how it was originally?

Comment (kphoenix137) [reply Source/engine/render/scrollrt.cpp:765]: Sideways walking also had its own unique handling for tile occupation, which was removed along with the southwards movement logic.

Comment (kphoenix137) [reply Source/engine/render/scrollrt.cpp:763]: The `WalkParameters` `startOffset` for `West` was the same as `SouthWest`, and `East` was the same was `SouthEast`.

In `getOffsetForWalking()`, it appears that moving `West` would shift the player to the right by `TILE_WIDTH` and moving `East` would shift the player to the left by `TILE_WIDTH`.

Comment (kphoenix137) [reply Source/engine/render/scrollrt.cpp:763]: Walking sideways also applied a manually adjustment of 1 tile on the positive y axis for `West` and 1 tile on the positive x axis for `East`. I'm really having a hard time making sense how it all worked, but I'm guessing it was just a hack for basically doing the same thing south walking was doing to a degree, where the player was being drawn as a part of a later drawn tile to get draw priority.

**Comment by kphoenix137** [comment]
Oh no it was merged already :D

**Comment by AJenbo** [comment]
@kphoenix137 set draft mode if you are still working on things...

**Comment by kphoenix137** [comment]
> @kphoenix137 set draft mode if you are still working on things...

I thought I was done. That's my bad lol. I was tinkering more with the PR and was like oh wow I think I fixed the other issue! 😆

**Comment by AJenbo** [comment]
arh no worries then :)

**Pull request merged**